File: src/airguitar/detectors/Piano.scala
package airguitar.detectors

import airguitar.Detector
import airguitar.Instrument
import airguitar.GestureIterator

import com.leapmotion.leap._
import com.leapmotion.leap.Gesture.State

class Piano(instrument: Instrument) extends Detector(instrument: Instrument) {

  override def onConnect(controller: Controller) {
    controller.enableGesture(Gesture.Type.TYPE_KEY_TAP)
  }

  override def onKeyTapGesture(gesture: KeyTapGesture) {
    val x: Float = gesture.position().getX()
    var n: Int   = ((((x + 120) * 50) / 240) + 50).toInt;
    super.getInstrument().play(n)
  }

}

File: src/airguitar/detectors/piano.clj
(ns airguitar.detectors.piano
  (:import [com.leapmotion.leap Controller Gesture Listener]
           [airguitar BasicInstrument GestureIterator])
  (:gen-class
    :name         airguitar.detectors.Piano
    :extends      [com.leapmotion.leap.Listener]
    :init         init
    :state        instrument))

(def key-tap-gesture
  (. com.leapmotion.leap.Gesture$Type TYPE_KEY_TAP))

(defn -init [^airguitar.BasicInstrument instrument]
  [[] instrument])

(defn -onConnect [this ^com.leapmotion.leap.Controller controller]
  (.enableGesture controller key-tap-gesture))

(defn -onFrame [this ^com.leapmotion.leap.Controller controller]
  (let [frame    (. controller frame)
        gestures (GestureIterator. (.getGestures frame))]
    (while (.hasNext gestures)
      (let [gesture (.next gestures)
            x       (int (.getX (.position gesture)))]
        (.play (.instrument this) x)))))
